test/e2e/alertmanager_test.go

```diff
@@ -472,32 +470,43 @@ inhibit_rules:
 		t.Fatal(err)
 	}
 
-	p := framework.MakeBasicPrometheus(ns, "test", "test", 3)
-	p.Spec.EvaluationInterval = "100ms"
-	framework.AddAlertingToPrometheus(p, ns, alertmanager.Name)
-
-	_, err = framework.MakeAndCreateFiringRule(ns, p.Name, alertName)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if err := framework.CreatePrometheusAndWaitUntilReady(ns, p); err != nil {
-		t.Fatal(err)
-	}
+	// Send alert to each Alertmanager
+	for i := 0; i < int(*alertmanager.Spec.Replicas); i++ {
+		replica := i
+		done := make(chan struct{})
+		errc := make(chan error, 1)
 
-	pSVC := framework.MakePrometheusService(p.Name, "not-relevant", v1.ServiceTypeClusterIP)
-	if finalizerFn, err := testFramework.CreateServiceAndWaitUntilReady(framework.KubeClient, ns, pSVC); err != nil {
-		t.Fatal(errors.Wrap(err, "creating Prometheus service failed"))
-	} else {
-		ctx.AddFinalizerFn(finalizerFn)
-	}
+		defer func() {
+			close(done)
+			select {
+			case err := <-errc:
+				t.Fatal(errors.Wrapf(err, "sending alert to alertmanager %v", replica))
+			default:
+				return
+			}
+		}()
+
+		go func() {
+			ticker := time.NewTicker(100 * time.Millisecond)
+			failures := 0
+			for {
+				select {
+				case <-ticker.C:
+					err := framework.SendAlertToAlertmanager(ns, "alertmanager-rolling-deploy-"+strconv.Itoa(replica))
+					if err != nil {
+						failures++
+						// Allow 30 (~3 Seconds) failures during Alertmanager rolling update.
```

Comment:
actually any amount of failures is fine as long as the result in the end is still correct

Reply:
The test can be split up into three phases:

1. Send alerts and expect a single notification.

2. Roll out new version of Alertmanager.

3. Send alerts and expect no new notifications.

If sending alerts between two and three fails, we will not see any new notifications, thereby the test might give a false positive.

Does that make sense @brancz?

Reply:
yep, makes perfect sense